File: src/main/java/core/procedures/cons/Car.java
package core.procedures.cons;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Cons;
import core.scm.Type;

import java.util.List;

public class Car extends AFn {

  public Car() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Type.Pair.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "car";
  }

  @Override
  public Object apply1(Object arg) {
    return car(arg);
  }

  public static Object car(Object o) {
    if (o instanceof Cons) {
      return ((Cons)o).car();
    }
    return ((List)o).get(0);
  }
}

File: src/main/java/core/procedures/cons/Car.kt
package core.procedures.cons

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Type
import core.utils.Utils

class Car : AFn<Any?, Any?>(name = "car", isPure = true, arity = Exactly(1),
                            mandatoryArgsTypes = arrayOf(Type.PairOrNonEmptyList::class.java)) {

    override operator fun invoke(arg: Any?): Any? {
        val seq = Utils.toSequence(arg)
        when (seq.none()) {
            true -> throw WrongTypeException(name, Type.PairOrNonEmptyList::class.java, emptyList<Nothing>())
            false -> return seq.first()
        }
    }
}
